### DISCUSSION TITLE
Disscussion Board

### DISCUSSION THREAD

**Discussion by vvaezian** [vvaezian/metabase_api_python#13] [General]
~~Topics that don't need a ticket, should be discussed here.~~
Please create a [new discussion](https://github.com/vvaezian/metabase_api_python/discussions/new/choose) if you have a question.

**Comment by santy1402** [comment]
Hello. I wanted to ask for your help with metabase_api. I searched the internet for a long time but could not find a solution. Perhaps you know. I have a local metabase and have google auth configured. How do I use the api in this case? I think that only you can help me. Thanks

**Comment by vvaezian** [comment]
Hi @santy1402 
I haven't tried Google authentication before. But probably one workaround is to use "ENABLE PASSWORD AUTHENTICATION" button. Then use username/password to connect to API (e.g. using `metabase_api_python` package, or other packages available).

![mb](https://user-images.githubusercontent.com/15974735/133149397-38fc7ccd-a2bb-44c6-9c78-bb44543a98e4.png)

**Reply by santy1402** [reply]
In this case, probably all users will be able to ignore the login through google. Maybe you know something else?

**Reply by vvaezian** [reply]
Cannot think of another approach at the moment.
Feel free to add the support for Google Auth to the current repo (see [here](https://github.com/vvaezian/metabase_api_python/blob/150c8143bf3ec964568d54bddd80bf9c1b2ca214/metabase_api/metabase_api.py#L17)).

**Reply by vvaezian** [reply]
Hi @santy1402,

At Intellimenta we have developed [DataBot](https://intellimenta.com/products/databot/), a conversational analytics app that connects with **Metabase** and enables users to ask questions about data and visualize the results using natural language in their preferred language.
If your company uses Metabase, DataBot can significantly enhance how your team and customers interact with data. 
If you're interested in learning more, you can book a demo using [this](https://outlook.office365.com/owa/calendar/booking@intellimenta.com/bookings/s/ZEMZuuM07U6C4Tr-GKUyOQ2) link. Happy to discuss further.

**Comment by AndrewLWilson** [comment]
@vvaezian I love what you've created here and would like to contribute if you are open to that!  Is there an easy way to populate a local Metabase instance with the required entities to run the test suite locally? I'm running the latest Metabase docker image, but the tests are failing because the required entities aren't found.

**Reply by vvaezian** [reply]
Hi @AndrewLWilson, I'm definitely open to any help for expanding the functionality of this package.
Regarding the tests, they are currently run against a live (personal) Metabase instance, so that wouldn't work on local. I cannot think of an easy way to maintain a local version currently. The tests will run once you create a PR though.

**Reply by AndrewLWilson** [reply]
Great!  For now I'll rely on the PR to run tests.  An idea to enable local testing for folks like me could be to redo the tests so that they bootstrap themselves off of the sample database that ships with Metabase.  The tests could start with creating a collection, then creating a card, a dashboard, etc.  That might be something I could work on if you are open to it.

**Reply by vvaezian** [reply]
That is a good idea. I'll look into it.

**Reply by vvaezian** [reply]
Feel free to write tests based on Sample Dataset. I had tried it initially, but the issue I believe was that I was missing some functions for example for creating dashboards. So I decided to use some pre-created objects.

**Reply by vvaezian** [reply]
@AndrewLWilson You can now run tests locally in v0.3.0 (see the [tests](https://github.com/vvaezian/metabase_api_python/tree/master/tests) directory).

**Reply by vvaezian** [reply]
Hi @AndrewLWilson,
At Intellimenta we have developed [DataBot](https://intellimenta.com/products/databot/), a conversational analytics app that connects with **Metabase** and enables users to ask questions about data and visualize the results using natural language in their preferred language.
If your company uses Metabase, DataBot can significantly enhance how your team and customers interact with data. 
If you're interested in learning more, you can book a demo using [this](https://outlook.office365.com/owa/calendar/booking@intellimenta.com/bookings/s/ZEMZuuM07U6C4Tr-GKUyOQ2) link. Happy to discuss further.

**Comment by KonstLofi** [comment]
@vvaezian Thanks a lot for you package! It very useful and helpful. Could you please explain why it is necessary to set `FRIENDLY TABLE AND FIELD NAMES` setting into `Disable` before use `get_columns_name_id` function?
The setting change only `display_name` property like `"display_name": "Goal Name"` and not change `name` property like `"name": "goal_name"`. I didn't find the use of `display_name` property in package.

**Reply by vvaezian** [reply]
Hi @KonstLofi , please see [this](https://github.com/metabase/metabase/issues/12419) issue.

**Reply by vvaezian** [reply]
Hi @KonstLofi,
At Intellimenta we have developed [DataBot](https://intellimenta.com/products/databot/), a conversational analytics app that connects with **Metabase** and enables users to ask questions about data and visualize the results using natural language in their preferred language.
If your company uses Metabase, DataBot can significantly enhance how your team and customers interact with data. 
If you're interested in learning more, you can book a demo using [this](https://outlook.office365.com/owa/calendar/booking@intellimenta.com/bookings/s/ZEMZuuM07U6C4Tr-GKUyOQ2) link. Happy to discuss further.

**Comment by KonstLofi** [comment]
I just comment the check for the setting and it works perfectly in my case.

**Reply by vvaezian** [reply]
Was the display name and real name of the field different in your example?

**Comment by KonstLofi** [comment]
Yes sir,

Have a look to attached json file.

**Reply by vvaezian** [reply]
I don't see the json file.
The disabling of `FRIENDLY TABLE AND FIELD NAMES` is required for the general case where the display name is different from the real name. As explained in the ticket mentioned above.

**Comment by KonstLofi** [comment]
{
  "description": null,
  "entity_type": "entity/GenericTable",
  "schema": "improvado_christianplanner",
  "db": {
    "description": null,
    "features": [],
    "cache_field_values_schedule": "0 0 4 * * ? *",
    "timezone": "UTC",
    "auto_run_queries": true,
    "metadata_sync_schedule": "0 42 * * * ? *",
    "name": "improvado_christianplanner",
    "settings": null,
    "caveats": null,
    "creator_id": 1,
    "is_full_sync": true,
    "updated_at": "2022-07-23T07:26:16.317",
    "cache_ttl": null,
    "details": null,
    "is_sample": false,
    "id": 40,
    "is_on_demand": false,
    "options": null,
    "engine": "bigquery-cloud-sdk",
    "initial_sync_status": "complete",
    "refingerprint": null,
    "created_at": "2022-07-23T07:15:18.607",
    "points_of_interest": null
  },
  "show_in_getting_started": false,
  "name": "activation_recurring_behaviour_report_christianplanner",
  "fields": [
    {
      "description": null,
      "database_type": "STRING",
      "semantic_type": "type/Category",
      "table_id": 2403,
      "coercion_strategy": null,
      "name": "goal_name",
      "fingerprint_version": 5,
      "has_field_values": "list",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [],
      "updated_at": "2022-07-24T18:43:11.306",
      "custom_position": 0,
      "effective_type": "type/Text",
      "active": true,
      "parent_id": null,
      "id": 85156,
      "last_analyzed": "2022-07-23T07:26:19.759",
      "position": 0,
      "visibility_type": "normal",
      "default_dimension_option": null,
      "target": null,
      "preview_display": true,
      "display_name": "Goal Name",
      "database_position": 0,
      "fingerprint": {
        "global": {
          "distinct-count": 4,
          "nil%": 0
        },
        "type": {
          "type/Text": {
            "percent-json": 0,
            "percent-url": 0,
            "percent-email": 0,
            "percent-state": 0,
            "average-length": 12.95
          }
        }
      },
      "created_at": "2022-07-23T07:15:49.145",
      "base_type": "type/Text",
      "points_of_interest": null
    },
    {
      "description": null,
      "database_type": "TIMESTAMP",
      "semantic_type": null,
      "table_id": 2403,
      "coercion_strategy": null,
      "name": "goal_date",
      "fingerprint_version": 5,
      "has_field_values": "none",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [
        "0",
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
        "7",
        "8",
        "9",
        "10",
        "11",
        "12",
        "13",
        "14"
      ],
      "updated_at": "2022-07-24T18:43:11.306",
      "custom_position": 0,
      "effective_type": "type/DateTimeWithLocalTZ",
      "active": true,
      "parent_id": null,
      "id": 85153,
      "last_analyzed": "2022-07-23T07:26:19.759",
      "position": 1,
      "visibility_type": "normal",
      "default_dimension_option": "2",
      "target": null,
      "preview_display": true,
      "display_name": "Goal Date",
      "database_position": 1,
      "fingerprint": {
        "global": {
          "distinct-count": 9591,
          "nil%": 0
        },
        "type": {
          "type/DateTime": {
            "earliest": "2022-01-21T12:23:37.943Z",
            "latest": "2022-07-22T03:06:21.432Z"
          }
        }
      },
      "created_at": "2022-07-23T07:15:49.144",
      "base_type": "type/DateTimeWithLocalTZ",
      "points_of_interest": null
    },
    {
      "description": null,
      "database_type": "STRING",
      "semantic_type": null,
      "table_id": 2403,
      "coercion_strategy": null,
      "name": "user_id",
      "fingerprint_version": 5,
      "has_field_values": "search",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [],
      "updated_at": "2022-07-24T18:43:11.306",
      "custom_position": 0,
      "effective_type": "type/Text",
      "active": true,
      "parent_id": null,
      "id": 85155,
      "last_analyzed": "2022-07-23T07:26:19.759",
      "position": 2,
      "visibility_type": "normal",
      "default_dimension_option": null,
      "target": null,
      "preview_display": true,
      "display_name": "User ID",
      "database_position": 2,
      "fingerprint": {
        "global": {
          "distinct-count": 8468,
          "nil%": 0
        },
        "type": {
          "type/Text": {
            "percent-json": 0,
            "percent-url": 0,
            "percent-email": 0,
            "percent-state": 0,
            "average-length": 35.8272
          }
        }
      },
      "created_at": "2022-07-23T07:15:49.145",
      "base_type": "type/Text",
      "points_of_interest": null
    },
    {
      "description": null,
      "database_type": "TIMESTAMP",
      "semantic_type": null,
      "table_id": 2403,
      "coercion_strategy": null,
      "name": "first_time_visit",
      "fingerprint_version": 5,
      "has_field_values": "none",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [
        "0",
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
        "7",
        "8",
        "9",
        "10",
        "11",
        "12",
        "13",
        "14"
      ],
      "updated_at": "2022-07-24T18:43:11.306",
      "custom_position": 0,
      "effective_type": "type/DateTimeWithLocalTZ",
      "active": true,
      "parent_id": null,
      "id": 85154,
      "last_analyzed": "2022-07-23T07:26:19.759",
      "position": 3,
      "visibility_type": "normal",
      "default_dimension_option": "2",
      "target": null,
      "preview_display": true,
      "display_name": "First Time Visit",
      "database_position": 3,
      "fingerprint": {
        "global": {
          "distinct-count": 8469,
          "nil%": 0
        },
        "type": {
          "type/DateTime": {
            "earliest": "2022-01-21T08:42:09.513Z",
            "latest": "2022-07-22T03:06:21.432Z"
          }
        }
      },
      "created_at": "2022-07-23T07:15:49.144",
      "base_type": "type/DateTimeWithLocalTZ",
      "points_of_interest": null
    }
  ],
  "caveats": null,
  "segments": [],
  "dimension_options": {},
  "updated_at": "2022-07-24T18:42:52.085",
  "active": true,
  "id": 2403,
  "db_id": 40,
  "visibility_type": null,
  "field_order": "database",
  "initial_sync_status": "complete",
  "display_name": "Activation Recurring Behaviour Report Christianplanner",
  "metrics": [],
  "created_at": "2022-07-23T07:15:22.748",
  "points_of_interest": null
}

**Comment by xandeea** [comment]
@vvaezian Hi, i'm getting an error using the function copy_dashboard: KeyError: 'sizeX', do you know what could be doing that?

**Reply by vvaezian** [reply]
hI @xandeea, it was due to a [change](https://github.com/vvaezian/metabase_api_python/issues/41) they made in Metabase API. I did release a patch for it. You just need to upgrade: `pip install --upgrade metabase_api`

**Reply by xandeea** [reply]
thanks a lot!

**Reply by vvaezian** [reply]
Hi @xandeea,
At Intellimenta we have developed [DataBot](https://intellimenta.com/products/databot/), a conversational analytics app that connects with **Metabase** and enables users to ask questions about data and visualize the results using natural language in their preferred language.
If your company uses Metabase, DataBot can significantly enhance how your team and customers interact with data. 
If you're interested in learning more, you can book a demo using [this](https://outlook.office365.com/owa/calendar/booking@intellimenta.com/bookings/s/ZEMZuuM07U6C4Tr-GKUyOQ2) link. Happy to discuss further.

**Comment by idot** [comment]
Dear @vvaezian , is it possible to use this metabase API client to migrate from one metabase instance to another? I saw some [projects](https://github.com/Credence-Analytics/metabase-cli) that do this, but I would start trying my migration with this project because it looks more complete/general if migration is in principle possible.

thank you very much in advance,
ido

**Reply by vvaezian** [reply]
Hi @idot, if you are using Pro or Enterprise version of Metabase, there is the [Serialization](https://www.metabase.com/docs/latest/installation-and-operation/serialization) feature specifically for your use-case. Otherwise, while you can use the current package for getting the Metabase objects, the current functions are not specifically designed for your use-case. So you probably would need to do some development yourself.

**Reply by vvaezian** [reply]
Hi @idot,
At Intellimenta we have developed [DataBot](https://intellimenta.com/products/databot/), a conversational analytics app that connects with **Metabase** and enables users to ask questions about data and visualize the results using natural language in their preferred language.
If your company uses Metabase, DataBot can significantly enhance how your team and customers interact with data. 
If you're interested in learning more, you can book a demo using [this](https://outlook.office365.com/owa/calendar/booking@intellimenta.com/bookings/s/ZEMZuuM07U6C4Tr-GKUyOQ2) link. Happy to discuss further.

**Comment by Umeshinduranga** [comment]
How can you abstract SQL writing for non devs?

**Reply by vvaezian** [reply]
Hi @Umeshinduranga,
Can you explain more?

**Reply by Umeshinduranga** [reply]
Most business users, analysts, or stakeholders can’t write SQL. But they know what they want. eg - any user can ask "show me sales by region","compare this month to last month". Therefor we can use predefined templates, dropdowns, or simple input forms .

**Reply by vvaezian** [reply]
Thanks for the clarification. Actually we have built a product called [DataBot](https://github.com/Intellimenta/DataBot) that does that. It allows users to ask data-related questions, make visualization, analyze data, generate forecast and more, all using natural language (in their native language).
It is **free**, **self-hosted**, and integrates with **Metabase**. You can try it out easily without installing anything. Instructions [here](https://databot.intellimenta.com/admin_guide/getting_started.html).
